src/util/safe_print.cpp

```diff
@@ -25,20 +28,20 @@
 namespace CVC4 {
 
 template <>
-void safe_print(int fd, const char* msg) {
-  ssize_t ret = write(fd, msg, safe_strlen(msg));
-  (void)ret;
-}
-
-template <>
-void safe_print(int fd, const std::string* msg) {
-  ssize_t ret = write(fd, msg->c_str(), msg->length());
-  (void)ret;
+void safe_print(int fd, const std::string& msg) {
+  // Print characters one by one instead of using
+  // string::data()/string::c_str() to avoid allocations (pre-c++11)
+  for (size_t i = 0; i < msg.length(); i++) {
+    if(write(fd, &(msg[i]), 1) != 1) {
+      abort();
+    }
+  }
 }
 
 template <>
-void safe_print(int fd, int64_t i) {
+void safe_print(int fd, const int64_t& _i) {
```

Comment:
Why a reference here? Not saying this is wrong, but I want to understand it. I am guessing there is a technical motivation. (And that might need to be documented.)